SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverGameplayDebuggerContext.cpp

```diff
@@ -0,0 +1,422 @@
+	RemovedNumber += ActorsAdded.Remove(EntityId);
+
+	if (RemovedNumber > 0)
+	{
+		bNeedToUpdateInterest = true;
+	}
+}
+
+void USpatialNetDriverGameplayDebuggerContext::OnComponentChange(Worker_EntityId EntityId, const SpatialGDK::ComponentChange& Change)
+{
+	if (Change.Type == SpatialGDK::ComponentChange::UPDATE)
+	{
+		ApplyComponentUpdate(EntityId, Change.Update);
+	}
+}
+
+void USpatialNetDriverGameplayDebuggerContext::ApplyComponentUpdate(Worker_EntityId EntityId, Schema_ComponentUpdate* Update)
+{
+	FEntityData* EntityData = TrackedEntities.Find(EntityId);
+	if (EntityData == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Warning, TEXT("Failed to ApplyComponentUpdate because entity isn't tracked"));
+		return;
+	}
+	EntityData->Component.ApplyComponentUpdate(Update);
+	bNeedToUpdateInterest = true;
+}
+
+SpatialGDK::QueryConstraint USpatialNetDriverGameplayDebuggerContext::ComputeAdditionalEntityQueryConstraint() const
+{
+	SpatialGDK::QueryConstraint EntitiesConstraint;
+	for (const auto& EntityData : TrackedEntities)
+	{
+		SpatialGDK::QueryConstraint EntityQuery;
+		EntityQuery.EntityIdConstraint = EntityData.Key;
+		EntitiesConstraint.OrConstraint.Add(EntityQuery);
+	}
+
+	return EntitiesConstraint;
+}
+
+void USpatialNetDriverGameplayDebuggerContext::ClearNeedEntityInterestUpdate()
+{
+	bNeedToUpdateInterest = false;
+}
+
+bool USpatialNetDriverGameplayDebuggerContext::NeedEntityInterestUpdate() const
+{
+	return bNeedToUpdateInterest;
+}
+
+void USpatialNetDriverGameplayDebuggerContext::OnServerRequest(AGameplayDebuggerCategoryReplicator* InCategoryReplicator,
+															   FString InServerWorkerId)
+{
+	check(NetDriver && NetDriver->PackageMap && NetDriver->Connection);
+
+	if (InCategoryReplicator == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Callback with null replicator"));
+		return;
+	}
+
+	Worker_EntityId EntityId = NetDriver->PackageMap->GetEntityIdFromObject(InCategoryReplicator);
+	if (EntityId == SpatialConstants::INVALID_ENTITY_ID)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Warning, TEXT("Callback from an actor with no entity?"));
+		return;
+	}
+
+	FEntityData* EntityData = TrackedEntities.Find(EntityId);
+	if (EntityData == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Warning, TEXT("Callback from an entity we are not tracking?"));
+		return;
+	}
+
+	if (EntityData->CurrentWorkerId != InServerWorkerId)
+	{
+		const uint32* VirtualWorkerId = PhysicalToVirtualWorkerIdMap.Find(InServerWorkerId);
+		if (VirtualWorkerId != nullptr)
+		{
+			EntityData->Component.DelegatedWorkerId = *VirtualWorkerId;
+			EntityData->CurrentWorkerId = InServerWorkerId;
+
+			InCategoryReplicator->SetCurrentServer(InServerWorkerId);
+
+			FWorkerComponentUpdate ComponentUpdate = EntityData->Component.CreateComponentUpdate();
+			NetDriver->Connection->SendComponentUpdate(EntityId, &ComponentUpdate);
```

Comment:
Can you collect the updates to be pushed later in TickServer instead ?

Reply:
aye I had it this way originally and then refactored it away, i'll be revert back to how it was.